LUCENE-9304: Refactor DWPTPool to pool DWPT directly

s1monw opened the pull request:
This change removes the `ThreadState` indirection from DWPTPool and pools DWPT directly. The tracking information and locking semantics are mostly moved to DWPT directly and the pool semantics have changed slightly such that DWPT need to be _checked-out_ in the pool once they need to be flushed or aborted. This automatically grows and shrinks the number of DWPT in the system when number of threads grow or shrink.  Access of pooled DWPTs is more straight forward and doesn't require ordinal. Instead consumers can just iterate over the elements in the pool. 
This allowed for removal of indirections in DWPTFlushControl like `BlockedFlush`, the removal of DWPTPool setter and getter in `IndexWriterConfig` and the addition of stronger assertions in DWPT and DW.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java:
```diff
@@ -322,35 +316,27 @@ synchronized Closeable lockAndAbortAll() throws IOException {
-      value = Math.max(value, perThread.lastSeqNo);
-    }
-    return value;
+    // NOCOMMIT: speak to mikemccandless about this change https://github.com/apache/lucene-solr/commit/5a03216/
```
Will do :)

s1monw replied:
👍

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java:
```diff
@@ -59,21 +58,25 @@
+ * that are associated with the current {@link DocumentsWriterDeleteQueue}
+ * out of the {@link DocumentsWriterPerThreadPool} and write
+ * them to disk. The flush process can piggy-back on incoming
+ * indexing threads or event block them from adding documents
```
s/`event`/`even`?

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java:
```diff
@@ -273,33 +269,31 @@ synchronized Closeable lockAndAbortAll() throws IOException {
-        perThreadPool.unlockNewThreadStates();
-        for (ThreadState state : threadStates) {
+        perThreadPool.unlockNewWriters();
+        for (DocumentsWriterPerThread state : writer) {
```
Isn't `writer` still an empty list here?  Oh I see, we are creating a lambda and only using it below, after we've added things to `writer`, OK.

dnhatn replied:
It might be less confusing if we reassign the `release` inside the `try` clause after acquiring DWPTs. Also nits: `writer` -> `writers` and `state` -> `writer`.

s1monw replied:
unfortunately we can't assign this later otherwise the ref can't be final. I will leave a comment.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java:
```diff
@@ -322,35 +316,27 @@ synchronized Closeable lockAndAbortAll() throws IOException {
-    return value;
+    // NOCOMMIT: speak to mikemccandless about this change https://github.com/apache/lucene-solr/commit/5a03216/
+    // Returning the last seqNum is as good as the way we had before IMO. I tried to figure out why this is better but
+    // failed.
```
I left a (non-review) comment on the single commit about this.

I think it's a good tradeoff?  Instead of making each indexing op compute `max` to save `lastSeqNo`, we are making caller of this API do the legwork.  I think it's a fair tradeoff since likely this API is very rarely called, but indexing ops are very often called.

s1monw replied:
hmm, I am not sure I understand fully. You mean the way it used to be was a good tradeoff? Currently there is no extra work done after my change, I am struggeling to understand why you made the chance in the first place. To me it's equivalent and no tests are failing.

mikemccand replied:
I am saying your change is the better tradeoff than the way it was before.

s1monw replied:
cool thanks for clarifying

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterFlushControl.java:
```diff
@@ -380,52 +368,35 @@ DocumentsWriterPerThread nextPendingFlush() {
+                return checkOutForFlush(next);
+              }
+            } finally {
+              next.unlock();
```
Is it intentional that we are `unlock`ing `next` even after returning it from `checkOutForFlush(next)`?  Shouldn't it remain locked until flush finishes?

s1monw replied:
so far it's intentional that we are checking out the DWPT of the pool and unlock it. The reason is that others might block on it to check it out as well but then won't succeed since it's not part of the pool anymore. I can take another iteration here and make sure it's locked until flushed. I wonder if we should do this after the fact.

mikemccand replied:
I was just worried that unlocking on return was not intended -- sometimes people think `finally` clauses are not executed on `return` ;)

If it is intentional that's great.

uschindler replied:
I think this should release again. The way of use of finally here is my favourite use of finally. You can make very elegant methods body's with returns and still have finally executed.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterFlushControl.java:
```diff
@@ -543,59 +506,53 @@ long markForFullFlush() {
       synchronized(this) {
-        if (!perThread.flushPending) {
-          setFlushPending(perThread);
+        if (dwpt.isFlushPending() == false) {
```
Woohoo!

s1monw replied:
<3

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterFlushControl.java:
```diff
@@ -717,51 +664,49 @@ public InfoStream getInfoStream() {
+      DocumentsWriterPerThread next = activePerThreadsIterator.next();
+      if (!next.isFlushPending()) {
+        final long nextRam = next.bytesUsed();
+        if (nextRam > 0 && next.getNumDocsInRAM() > 0) {
```
Hmm why would we ever have `nextRam > 0` but `next.getNumDocsInRAM() == `?

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThread.java:
```diff
@@ -600,5 +608,81 @@ public String toString() {
+  }
+
+  /**
+   * Acquires the DWPTs lock only if it is not held by another thread at the time
```
s/`DWPTs`/`DPWT's`

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThread.java:
```diff
@@ -600,5 +608,81 @@ public String toString() {
+  }
+
+  /**
+   * Returns true if the DWPTs lock is held by the current thread
```
s/`DWPTs`/`DWPT's`

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThread.java:
```diff
@@ -600,5 +608,81 @@ public String toString() {
+  }
+
+  /**
+   * Unlocks the DWPTs lock
```
Here too.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThreadPool.java:
```diff
@@ -16,228 +16,173 @@
-   * before accessing the state.
-   */
-  @SuppressWarnings("serial")
-  final static class ThreadState extends ReentrantLock {
```
Woo hoo!  I always thought it was kinda spooky that we were subclassing `ReentrantLock` instead of "hasa" that this change is switching to!

s1monw replied:
yeah ++

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThread.java:
```diff
@@ -156,6 +158,9 @@ void abort() throws IOException{
   final SegmentInfo segmentInfo;     // Current segment we are working on
   private boolean aborted = false;   // True if we aborted
+  private SetOnce<Boolean> flushPending = new SetOnce<>();
+  private volatile long lastCommittedBytesUsed = 0;
```
`= 0` isn't needed -- it's Java's default already.

s1monw replied:
👍

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterFlushControl.java:
```diff
@@ -458,74 +429,66 @@ public void setApplyAllDeletes() {
+          addFlushableDWPT(perThread);
+        } else {
+          unlock = false;
+          // simply return the DWPT even in a flush all case sine we already hold the lock
```
s/`sine`/`since`

dnhatn commented on lucene/core/src/java/org/apache/lucene/index/FlushPolicy.java:
```diff
@@ -52,38 +50,38 @@
    */
   public abstract void onDelete(DocumentsWriterFlushControl control,
-      ThreadState state);
+                                DocumentsWriterPerThread state);
```
state -> dwpt or perThread?

s1monw replied:
👍

s1monw commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:
```diff
@@ -3774,6 +3768,9 @@ public void testRefreshAndRollbackConcurrently() throws Exception {
       indexer.join();
       refresher.join();
+      if (w.getTragicException() != null) {
+        w.getTragicException().printStackTrace();
```
yeah I ran into this multiple times. I fixed it for real this time.

s1monw commented:
> Have we changed the behavior of thread affinity (which DWPT an indexing thread will pull to index a new document), or of flushing (e.g., we still hard stall incoming indexing threads if flushing cannot keep up due to slow IO device)? It looks like we have not but I wanted to confirm.

no we haven't. I changed the locking semantics slightly in DWPTPool to move to tryLock when we get a new DWPT but that shouldn't make any difference in terms of thread affinity.

s1monw commented:
@mikemccand @dnhatn I pushed new commits and replied to you comments

s1monw commented:
@mikemccand @dnhatn I went through DWPTFlushControl one more time to add documentation to all the different lists we have in there and cleaned up some more stuff which makes it a bit less confusing. Not many changes but another pass would be great.

dnhatn commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterFlushControl.java:
```diff
@@ -52,12 +52,19 @@
   // only for safety reasons if a DWPT is close to the RAM limit
   private final Queue<DocumentsWriterPerThread> blockedFlushes = new LinkedList<>();
+  // flushingWriters holds all currently flushing writers. There might be writers in this list that
+  // are also in the flushQueue which means that writers in teh flushingWriters list are not necessarily
```
nit: teh -> the

uschindler commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThreadPool.java:
```diff
@@ -16,228 +16,173 @@
 
-  private final List<ThreadState> threadStates = new ArrayList<>();
+  private final Set<DocumentsWriterPerThread> dwpts = Collections.newSetFromMap(new IdentityHashMap<>());
+  private final List<DocumentsWriterPerThread> freeList = new ArrayList<>();
```
Just an idea: Maybe use `Deque<DocumentsWriterPerThread> freeList = new ArrayDeque<>();` here, as it allows to iterate in borth directions (it has `descendingIterator()`). To me this also looks better, because the whole thing is mostly used as a deque (LIFO).

s1monw replied:
👍

uschindler commented on lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThreadPool.java:
```diff
@@ -16,228 +16,173 @@
+      // Important that we are LIFO here! This way if number of concurrent indexing threads was once high,
+      // but has now reduced, we only use a limited number of DWPTs. This also guarantees that if we have suddenly
+      // a single thread indexing
+      for (int i = freeList.size()-1; i >= 0; i--) {
```
Here you could use the `descendingIterator()` and iterate while just calling remove() on the iterator.

s1monw replied:
👍

s1monw commented:
@uschindler I pushed feedback - thanks for taking a look

uschindler commented:
> @uschindler I pushed feedback - thanks for taking a look

No problem. I have just seen the index based backwards iteration, which is fine for random access lists.

`ArrayDeque` is often forgotten, although is much better than lists for some cases (but not general replacement). Whenever you need to insert/remove elements on both sides of the list it's no lineral operation (especially inserting removing elements at begin). Behaves like `LinkedList` but with better CPU cache semantics.

mikemccand commented:
I'll look again -- in the meantime I beasted all Lucene (core + modules) tests and hit this failure.  It does not reproduce, and I doubt it's related to / caused by this change, since it looks like it's a file ref counting issue on exception:

```
org.apache.lucene.index.TestAddIndexes > test suite's output saved to /home/mike/src/simon/lucene/core/build/test-results/test/outputs/OUTPUT-org.apache.lucene.index.TestAddIndexes.txt, copied below:
   >     java.nio.file.NoSuchFileException: _gt_Lucene85FieldsIndex-doc_ids_6u.tmp
   >         at __randomizedtesting.SeedInfo.seed([4760FA81FBD4B2CE:A147156E5F7BD9B0]:0)
   >         at org.apache.lucene.store.ByteBuffersDirectory.deleteFile(ByteBuffersDirectory.java:148)
   >         at org.apache.lucene.store.MockDirectoryWrapper.deleteFile(MockDirectoryWrapper.java:607)
   >         at org.apache.lucene.store.LockValidatingDirectoryWrapper.deleteFile(LockValidatingDirectoryWrapper.java:38)
   >         at org.apache.lucene.index.IndexFileDeleter.deleteFile(IndexFileDeleter.java:696)
   >         at org.apache.lucene.index.IndexFileDeleter.deleteFiles(IndexFileDeleter.java:690)
   >         at org.apache.lucene.index.IndexFileDeleter.refresh(IndexFileDeleter.java:449)
   >         at org.apache.lucene.index.IndexWriter.rollbackInternalNoCommit(IndexWriter.java:2334)
   >         at org.apache.lucene.index.IndexWriter.rollbackInternal(IndexWriter.java:2275)
   >         at org.apache.lucene.index.IndexWriter.rollback(IndexWriter.java:2268)
   >         at org.apache.lucene.index.TestAddIndexes.testAddIndexesWithRollback(TestAddIndexes.java:974)
   >         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   >         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   >         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   >         at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:942)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:978)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:992)
   >         at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
   >         at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
   >         at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
   >         at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
   >         at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:819)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:470)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:951)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:836)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:887)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:898)
   >         at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
   >         at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
   >         at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
   >         at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:826)
   >         at java.base/java.lang.Thread.run(Thread.java:834)
  2> NOTE: reproduce with: ant test  -Dtestcase=TestAddIndexes -Dtests.method=testAddIndexesWithRollback -Dtests.seed=4760FA81FBD4B2CE -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=fr-GP -Dtests.t\
imezone=Asia/Tbilisi -Dtests.asserts=true -Dtests.file.encoding=UTF-8
  2> NOTE: test params are: codec=Asserting(Lucene84): {c=PostingsFormat(name=LuceneFixedGap), id=PostingsFormat(name=LuceneFixedGap), f1=PostingsFormat(name=LuceneFixedGap), f2=BlockTreeOrds(blocksize=128)\
, version=BlockTreeOrds(blocksize=128), content=FST50}, docValues:{dv=DocValuesFormat(name=Lucene80), soft_delete=DocValuesFormat(name=Lucene80), doc=DocValuesFormat(name=Lucene80), id=DocValuesFormat(name=\
Asserting), content=DocValuesFormat(name=Asserting), doc2d=DocValuesFormat(name=Lucene80)}, maxPointsInLeafNode=982, maxMBSortInHeap=5.837219998050092, sim=Asserting(org.apache.lucene.search.similarities.As\
sertingSimilarity@6ce38471), locale=fr-GP, timezone=Asia/Tbilisi
  2> NOTE: Linux 5.5.6-arch1-1 amd64/Oracle Corporation 11.0.6 (64-bit)/cpus=128,threads=1,free=120119504,total=268435456
  2> NOTE: All tests run in this JVM: [TestXYPoint, TestPoint, TestLucene80NormsFormat, TestHighCompressionMode, TestPayloadsOnVectors, TestTermVectorsWriter, TestAddIndexes]
```

dnhatn commented:
> It does not reproduce, and I doubt it's related to / caused by this change, since it looks like it's a file ref counting issue on exception:

Thanks @mikemccand. I think the failure is not related to this PR (but we should look into it). I saw the same test failure on Elastic CI two days ago.

s1monw commented:
> I'll look again -- in the meantime I beasted all Lucene (core + modules) tests and hit this failure. It does not reproduce, and I doubt it's related to / caused by this change, since it looks like it's a file ref counting issue on exception:

yeah it doesn't look related. I will still look into it unless @dnhatn beats me

s1monw commented:
I opened [LUCENE-9309](https://issues.apache.org/jira/browse/LUCENE-9309) for this

mikemccand commented:
Hmm here's a more exotic test failure, also likely not caused by the changes here.  The fun things you learn when beasting on a 128 core box :)

Hmm, though it is remotely possible my storage device or ECC RAM is flipping bits:

```
org.apache.lucene.index.TestIndexManyDocuments > test suite's output saved to /home/mike/src/simon/lucene/core/build/test-results/test/outputs/OUTPUT-org.apache.lucene.index.TestIndexManyDocuments.txt, copi\
ed below:
  1> CheckIndex failed
  1> 0.00% total deletions; 10976 documents; 0 deleteions
  1> Segments file=segments_1 numSegments=4 version=9.0.0 id=cekay2d5izae12ssuqikoqgoc
  1>   1 of 4: name=_d maxDoc=8700
  1>     version=9.0.0
  1>     id=cekay2d5izae12ssuqikoqgob
  1>     codec=Asserting(Lucene84)
  1>     compound=false
  1>     numFiles=11
  1>     size (MB)=0.003
  1>     diagnostics = {os.version=5.5.6-arch1-1, java.vendor=Oracle Corporation, source=merge, os.arch=amd64, mergeFactor=10, java.runtime.version=11.0.6+8-LTS, os=Linux, timestamp=1586347074798, lucene.ve\
rsion=9.0.0, java.vm.version=11.0.6+8-LTS, java.version=11.0.6, mergeMaxNumSegments=-1}
  1>     no deletions
  1>     test: open reader.........OK [took 0.001 sec]
  1>     test: check integrity.....OK [took 0.000 sec]
  1>     test: check live docs.....OK [took 0.000 sec]
  1>     test: field infos.........OK [1 fields] [took 0.000 sec]
  1>     test: field norms.........OK [1 fields] [took 0.001 sec]
  1>     test: terms, freq, prox...ERROR: java.lang.AssertionError: buffer=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0] bufferSize=1024 buffer.length=0
  1> java.lang.AssertionError: buffer=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0] bufferSize=1024 buffer.length=0
  1>    at org.apache.lucene.store.BufferedIndexInput.setBufferSize(BufferedIndexInput.java:78)
  1>    at org.apache.lucene.codecs.MultiLevelSkipListReader.loadSkipLevels(MultiLevelSkipListReader.java:241)
  1>    at org.apache.lucene.codecs.MultiLevelSkipListReader.init(MultiLevelSkipListReader.java:208)
  1>    at org.apache.lucene.codecs.lucene84.Lucene84SkipReader.init(Lucene84SkipReader.java:103)
  1>    at org.apache.lucene.codecs.lucene84.Lucene84PostingsReader$EverythingEnum.advance(Lucene84PostingsReader.java:837)
  1>    at org.apache.lucene.index.FilterLeafReader$FilterPostingsEnum.advance(FilterLeafReader.java:271)
  1>    at org.apache.lucene.index.AssertingLeafReader$AssertingPostingsEnum.advance(AssertingLeafReader.java:377)
  1>    at org.apache.lucene.index.CheckIndex.checkFields(CheckIndex.java:1426)
  1>    at org.apache.lucene.index.CheckIndex.testPostings(CheckIndex.java:1867)
  1>    at org.apache.lucene.index.CheckIndex.checkIndex(CheckIndex.java:720)
  1>    at org.apache.lucene.util.TestUtil.checkIndex(TestUtil.java:301)
  1>    at org.apache.lucene.util.TestUtil.checkIndex(TestUtil.java:286)
  1>    at org.apache.lucene.store.BaseDirectoryWrapper.close(BaseDirectoryWrapper.java:45)
  1>    at org.apache.lucene.util.IOUtils.close(IOUtils.java:89)
  1>    at org.apache.lucene.util.IOUtils.close(IOUtils.java:77)
  1>    at org.apache.lucene.index.TestIndexManyDocuments.test(TestIndexManyDocuments.java:69)
  1>    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  1>    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  1>    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  1>    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:942)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:978)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:992)
  1>    at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  1>    at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  1>    at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  1>    at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  1>    at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  1>    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  1>    at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
  1>    at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:819)
  1>    at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:470)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:951)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:836)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:887)
  1>    at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:898)
  1>    at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  1>    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  1>    at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
  1>    at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  1>    at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  1>    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  1>    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  1>    at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  1>    at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  1>    at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  1>    at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  1>    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  1>    at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
  1>    at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:826)
  1>    at java.base/java.lang.Thread.run(Thread.java:834)
  1>     test: stored fields.......OK [0 total field count; avg 0.0 fields per doc] [took 0.001 sec]
  1>     test: term vectors........OK [0 total term vector count; avg 0.0 term/freq vector fields per doc] [took 0.000 sec]
  1>     test: docvalues...........OK [0 docvalues fields; 0 BINARY; 0 NUMERIC; 0 SORTED; 0 SORTED_NUMERIC; 0 SORTED_SET] [took 0.000 sec]
  1>     test: points..............OK [0 fields, 0 points] [took 0.000 sec]
  1> FAILED
  1>     WARNING: exorciseIndex() would remove reference to this segment; full exception:
  1> java.lang.RuntimeException: Term Index test failed
  1> WARNING: 1 broken segments (containing 8700 documents) detected
  1> Took 0.013 sec total.
  1>
   >     java.lang.RuntimeException: CheckIndex failed
   >         at __randomizedtesting.SeedInfo.seed([CB92BC7F7BCB5A04:43C683A5D53737FC]:0)
   >         at org.apache.lucene.util.TestUtil.checkIndex(TestUtil.java:306)
   >         at org.apache.lucene.util.TestUtil.checkIndex(TestUtil.java:286)
   >         at org.apache.lucene.store.BaseDirectoryWrapper.close(BaseDirectoryWrapper.java:45)
   >         at org.apache.lucene.util.IOUtils.close(IOUtils.java:89)
   >         at org.apache.lucene.util.IOUtils.close(IOUtils.java:77)
   >         at org.apache.lucene.index.TestIndexManyDocuments.test(TestIndexManyDocuments.java:69)
   >         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   >         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   >         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   >         at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:942)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:978)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:992)
   >         at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
   >         at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
   >         at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
   >         at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
   >         at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:819)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:470)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:951)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:836)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:887)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:898)
   >         at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
   >         at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
   >         at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
   >         at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:826)
   >         at java.base/java.lang.Thread.run(Thread.java:834)
  2> NOTE: reproduce with: ant test  -Dtestcase=TestIndexManyDocuments -Dtests.method=test -Dtests.seed=CB92BC7F7BCB5A04 -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=sr-Cyrl-ME -Dtests.timezone=A\
merica/Los_Angeles -Dtests.asserts=true -Dtests.file.encoding=UTF-8
  2> NOTE: leaving temporary files on disk at: /home/mike/src/simon/lucene/core/build/tmp/tests-tmp/lucene.index.TestIndexManyDocuments_CB92BC7F7BCB5A04-001
  2> NOTE: test params are: codec=Asserting(Lucene84), sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@4fe8fb1e), locale=sr-Cyrl-ME, timezone=America/Los_Angeles
  2> NOTE: Linux 5.5.6-arch1-1 amd64/Oracle Corporation 11.0.6 (64-bit)/cpus=128,threads=1,free=249224512,total=268435456
  2> NOTE: All tests run in this JVM: [TestSearchForDuplicates, TestCircle2D, TestLucene80NormsFormatMergeInstance, TestFastDecompressionMode, TestMultiTermsEnum, TestFieldInvertState, TestFieldReuse, TestC\
oncurrentMergeScheduler, TestTerms, TestFieldUpdatesBuffer, TestIndexWriterFromReader, TestReadOnlyIndex, TestTragicIndexWriterDeadlock, TestIndexManyDocuments]

```

uschindler commented:
This could be related to Adrien's changes yesterday: https://issues.apache.org/jira/browse/issue/LUCENE-9271

s1monw commented:
@mikemccand did you run any benchmarks on this change yet?

jpountz commented:
@mikemccand I think it's related to the change I merged yesterday indeed. I fix it shortly after merging, so if you merge master back, this should address the failure.

s1monw commented:
@mikemccand I merged master into this branch

mikemccand commented:
I will try to run benchmarks soon now, sorry for the delay.  We are furiously 3D printing / laser cutting face shields for local hospitals ...

uschindler commented:
> We are furiously 3D printing / laser cutting face shields for local hospitals ...

You need all CPU power for this right? 🤪

s1monw commented:
> You need all CPU power for this right? 🤪

I assume this time the limited resource are the hands using the keyboard :)

mikemccand commented:
OK indexing throughput is noisy, but net/net I think there's no perf impact.  This is `wikimediumall` from `luceneutil` on 128 core box, JDK 11:

```
/l/logs/simon1.log:Indexer: 223.42653129383794 GB/hour plain text
/l/logs/simon2.log:Indexer: 264.0324599263973 GB/hour plain text
/l/logs/simon3.log:Indexer: 215.7334958611665 GB/hour plain text

/l/logs/trunk1.log:Indexer: 209.16003685553957 GB/hour plain text
/l/logs/trunk2.log:Indexer: 210.4946120847491 GB/hour plain text
/l/logs/trunk3.log:Indexer: 231.3400394444364 GB/hour plain text
```

mikemccand commented:
> I assume this time the limited resource are the hands using the keyboard :)

Indeed!

s1monw commented:
@mikemccand I will go ahead and merge this! thanks!

mikemccand commented:
++!
